### PR TITLE
feat(x/warden): Add metrics/events to x/warden and x/intent

### LINKED ISSUE #222: Add metrics/events to x/warden and x/intent
## Events

> Events are objects that contain information about the execution of the application. They are mainly used by service providers like block explorers and wallet to track the execution of various messages and index transactions.
> https://docs.cosmos.network/v0.50/learn/advanced/events

Each event should be a protobuf message (let's create a `events.proto` for each module), then can be emitted during transaction handling as described here:

See https://docs.cosmos.network/v0.50/build/building-modules/msg-services#events for adding new events.

## Telemetry

> Gather relevant insights about your application and modules with custom metrics and telemetry.
> https://docs.cosmos.network/v0.50/learn/advanced/telemetry

Events are for indexing transactions, metrics are for monitoring performances. Sometimes the two can overlap and that's why I'm putting them together in a single issue.

See https://docs.cosmos.network/v0.50/build/building-modules/msg-services#telemetry for adding new metrics.

## What to track

This is the main question. For an indexer, we can have some interesting analytics events, a non-exhaustive list could be:

- space created
- space owner added/removed
- key requested
- key fulfilled/rejected
- signature requested
- signature fulfilled/rejected
- intent created
- action created
- action executed
- action expired

This would let an indexer be able to build custom pages, for example, to list all transactions related to a specific Action, a specific Key, etc. We should also include the referenced accounts, so that the indexer can also have a page for a specific account showing all its Spaces, Keys, etc.

For metrics, I would probably be interested in counting how many times a Msg is being executed and the time spent handling it, so we can view bottlenecks in our system. E.g. something like this: https://github.com/cosmos/cosmos-sdk/blob/00ec23cdbe1b2bef180021e4e4ac569e6f26bcae/x/staking/keeper/abci.go#L15 but for single msg handlers instead of just BeginBlocker.

### PR DESCRIPTION
**Pull request by Overseven**
closes #222

PR contains events only. Metrics will be added later in this PR

I created these events using three principles:
1. do not use arrays and strings with unpredictable length
2. do not duplicate data if it is not changed 
3. use a hash instead of raw calldata

# Events
## Warden module events
Related to Spaces:
- `EventCreateSpace(id, creator, owners_count, admin_intent_id, sign_intent_id)`
- `EventUpdateSpace(space_id, admin_intent_id, sign_intent_id)`
- `EventAddSpaceOwner(space_id, new_owner, owners_count)`
- `EventRemoveSpaceOwner(space_id, removed_owner, owners_count)`

Related to Keys:
- `EventRequestNewKey(id, space_id, keychain_id, intent_id, key_type, creator)`
- `EventRequestNewKeyFulfilled(request_id, public_key)`
- `EventRequestNewKeyRejected(request_id)`
- `EventUpdateKey(key_id, intent_id)`

Related to Signatures:
- `EventRequestNewSignature(id, key_id, signing_data_hash, creator)`
- `EventRequestNewSignatureFulfilled(request_id)`
- `EventRequestNewSignatureRejected(request_id)`

Related to Keychains:
- `EventNewKeychain(id, creator, admin_intent_id, keychain_fees)`
- `EventUpdateKeychain(id, keychain_fees, is_active)`
- `EventAddKeychainParty(id, new_party, party_count)`

## Intent module events
- `EventCreateIntent(id, creator)`
- `EventUpdateIntent(id)`

# Metrics
Action execution time measurement was added.

List of new metrics (counters):
- `action_count`
- `approve_action.msg.count`
- `revoke_action.msg.count`
- `new_intent.msg.count`
- `update_intent.msg.count`
- `new_keychain.msg.count`
- `add_keychain_party.msg.count`
- `update_keychain.msg.count`
- `new_space.msg.count`
- `update_space.msg.count`
- `add_space_owner.msg.count`
- `remove_space_owner.msg.count`
- `new_key_request.msg.count`
- `update_key_request.msg.count`
- `update_key.msg.count`
- `new_signature_request.msg.count`
- `fulfill_signature_request.msg.count`
- `warden.update_params.msg.count`
- `intent.update_params.msg.count`

## Summary by CodeRabbit

- **New Features**
	- Introduced telemetry tracking across various functions to monitor and measure performance and event handling.
	- Added new event types for tracking intent and key management within the `x/warden` and `x/intent` modules.

- **Documentation**
	- Updated `CHANGELOG.md` to reflect new events and telemetry metrics for the `x/intent` and `x/warden` modules.

### CODE REVIEW

File: warden/x/intent/keeper/msg_server_new_intent.go [hunk 1]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/warden-protocol/wardenprotocol/api/warden/intent"
```
Comment (pitasi) [inline warden/x/intent/keeper/msg_server_new_intent.go]: don't import the API pkg from here, the correct "types" package is already imported, use that

it also applies to all the other files

File: proto/warden/intent/events.proto [hunk 2]
```diff
@@ -0,0 +1,20 @@
+syntax = "proto3";
+
+package warden.intent;
+
+option go_package = "github.com/warden-protocol/wardenprotocol/warden/x/intent/types";
+
+// EventCreateIntent is emitted on Intent creation
+message EventCreateIntent {
+  // intent_id associated with the intent.
+  uint64 intent_id = 1;
```
Comment (pitasi) [inline proto/warden/intent/events.proto]:
```suggestion
  // id of the new Intent.
  uint64 id = 1;
```

so we maintain consistency with the actual message fields

File: proto/warden/intent/events.proto [hunk 3]
```diff
@@ -0,0 +1,20 @@
+syntax = "proto3";
+
+package warden.intent;
+
+option go_package = "github.com/warden-protocol/wardenprotocol/warden/x/intent/types";
+
+// EventCreateIntent is emitted on Intent creation
```
Comment (pitasi) [inline proto/warden/intent/events.proto:9]:
```suggestion
// EventCreateIntent is emitted on Intent creation.
```

🕵️‍♂️ <sup>I saw you put a period in the next comments, so let's add it to all of them</sup>

File: proto/warden/intent/events.proto [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+message EventCreateIntent {
+  // intent_id associated with the intent.
+  uint64 intent_id = 1;
+
+  // creator is the account address of the intent creator
+  string creator = 2;
+}
+
+// EventUpdateIntent is emitted on Intent updating
```
Comment (pitasi) [inline proto/warden/intent/events.proto]:
```suggestion
// EventUpdateIntent is emitted when Intent is updated.
```

File: proto/warden/warden/v1beta2/events.proto [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+// EventCreateSpace is emitted on Space creation
+message EventCreateSpace {
+  // space_id associated with the space.
+  uint64 space_id = 1;
+
+  // creator is the account address of the space creator.
+  string creator = 2;
+
+  // owners_count is the count of the space owners.
+  uint64 owners_count = 3;
```
Comment (pitasi) [inline proto/warden/warden/v1beta2/events.proto]: it's fine tracking this number, but then I expect it to be tracked every time this count changes as well (so in AddSpaceOwner and RemoveSpaceOwner) :)

File: warden/x/intent/keeper/msg_server_new_intent.go [hunk 6]
```diff
@@ -27,6 +28,16 @@ func (k msgServer) NewIntent(goCtx context.Context, msg *types.MsgNewIntent) (*t
...
+	err = k.EventService.EventManager(ctx).Emit(ctx,
+		&intent.EventCreateIntent{
+			IntentId: id,
+			Creator:  msg.Creator,
+		})
+
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (pitasi) [inline warden/x/intent/keeper/msg_server_new_intent.go]:
```suggestion
	sdkCtx := sdk.UnwrapSDKContext(ctx)
	err = sdkCtx.EventManager().EmitTypedEvent(
		&intent.EventCreateIntent{
			IntentId: id,
			Creator:  msg.Creator,
		})
	if err != nil {
		return nil, err
	}
```

I think that this is the correct way of accessing the EventManager from inside a MsgHandler (but I haven't tested it myself)

**Review by pitasi — CHANGES_REQUESTED** [review]

File: warden/x/intent/keeper/msg_server_new_intent.go (hunk 1 above)
Comment (Overseven) [reply warden/x/intent/keeper/msg_server_new_intent.go]: The redundant imports have been removed

File: proto/warden/intent/events.proto (hunk 2 above)
Comment (Overseven) [reply proto/warden/intent/events.proto]: I have renamed the fields according to the messages

File: proto/warden/intent/events.proto (hunk 3 above)
Comment (Overseven) [reply proto/warden/intent/events.proto:9]: fixed

File: proto/warden/intent/events.proto (hunk 4 above)
Comment (Overseven) [reply proto/warden/intent/events.proto]: fixed

File: proto/warden/warden/v1beta2/events.proto (hunk 5 above)
Comment (Overseven) [reply proto/warden/warden/v1beta2/events.proto]: The same field for these events was added

File: warden/x/intent/keeper/msg_server_new_intent.go (hunk 6 above)
Comment (Overseven) [reply warden/x/intent/keeper/msg_server_new_intent.go]: I will test both options on localnet

**Comment by pitasi** [comment]
superseeded by #410 #411 #441

**Pull request closed without merging**